### PR TITLE
Set node providerID on target cluster

### PR DESCRIPTION
**Pull request by maelk**
When no cloud provider is deployed, set the providerID by
directly interacting with the target cluster api.

### CODE REVIEW

File: baremetal/remote/remote_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+var (
+	clusterWithValidKubeConfig = &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test1",
+			Namespace: "test",
+		},
+	}
+
+	clusterWithInvalidKubeConfig = &clusterv1.Cluster{
```
Comment (Jaakko-Os) [inline baremetal/remote/remote_test.go]: what makes the config invalid?
maybe Name: "test2-invalid" ?

File: baremetal/remote/remote_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	clusterWithNoKubeConfig = &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test3",
+			Namespace: "test",
+		},
+	}
+
+	validKubeConfig = `
```
Comment (Jaakko-Os) [inline baremetal/remote/remote_test.go]: alignment ?

File: baremetal/remote/remote_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			Name:      "test1-kubeconfig",
+			Namespace: "test",
+		},
+		Data: map[string][]byte{
+			secret.KubeconfigDataName: []byte(validKubeConfig),
+		},
+	}
+
+	invalidSecret = &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
```
Comment (Jaakko-Os) [inline baremetal/remote/remote_test.go]: similar issue here, how to know that secret is invalid?

File: controllers/baremetalmachine_controller_test.go [hunk 4]
```diff
@@ -325,6 +336,164 @@ var _ = Describe("Reconcile Baremetalcluster", func() {
...
+								APIVersion: infrav1.GroupVersion.String(),
+							},
+							Bootstrap: clusterv1.Bootstrap{
+								Data: &bootstrapData,
+							},
+						},
+						Status: clusterv1.MachineStatus{
+							BootstrapReady: true,
+						},
+					},
```
Comment (Jaakko-Os) [inline controllers/baremetalmachine_controller_test.go]: the above is repeated below, could be defined once and reused for better readability

File: baremetal/remote/remote_test.go (hunk 2 above)
Comment (maelk) [reply baremetal/remote/remote_test.go]: alignment is correct as far as I can tell

File: baremetal/baremetalmachine_manager.go [hunk 5]
```diff
@@ -654,3 +646,38 @@ func (m *MachineManager) nodeAddresses(host *bmh.BareMetalHost) []capi.MachineAd
...
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	for _, node := range nodes.Items {
+		if node.Spec.ProviderID == providerID {
+			continue
+		}
+		node.Spec.ProviderID = providerID
+		_, err = corev1Remote.Nodes().Update(&node)
+		if err != nil {
+			return errors.Wrap(err, "unable to get update node for baremetal host")
```
Comment (Xenwar) [inline baremetal/baremetalmachine_manager.go]: unable to update

File: baremetal/remote/remote_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		if !strings.Contains(err.Error(), "not found") {
+			t.Fatalf("Expected not found error, got %v", err)
+		}
+	})
+
+	t.Run("cluster with invalid kubeconfig", func(t *testing.T) {
+		client := fake.NewFakeClient(invalidSecret)
+		_, err := NewClusterClient(client, clusterWithInvalidKubeConfig)
+		if err == nil || apierrors.IsNotFound(err) {
+			t.Fatalf("Expected error, got %v", err)
```
Comment (kashifest) [inline baremetal/remote/remote_test.go]: if its an apierrors.IsNotFound(err), then the message will be, 'Expected error, got not found error'. Should we rephrase it?

File: controllers/baremetalmachine_controller_test.go [hunk 7]
```diff
@@ -511,6 +522,176 @@ var _ = Describe("Reconcile Baremetalcluster", func() {
...
+			},
+		),
+		//Given: Baremetalmachine with annotation to a BMH provisioned, machine with
+		// bootstrap data, target cluster node available
+		//Expected: no error, no requeue. ProviderID should be set.
+		Entry("Should not requeue when patching an available node",
+			TestCaseReconcile{
+				Objects: []runtime.Object{
+					&infrav1.BareMetalMachine{
+						TypeMeta: metav1.TypeMeta{},
```
Comment (kashifest) [inline controllers/baremetalmachine_controller_test.go]: Kind is missing?

File: controllers/baremetalmachine_controller_test.go [hunk 8]
```diff
@@ -511,6 +522,176 @@ var _ = Describe("Reconcile Baremetalcluster", func() {
...
 			},
 		),
+		//Given: Baremetalmachine with annotation to a BMH provisioned, machine with
+		// bootstrap data, no target cluster node available
+		//Expected: no error, requeing. ProviderID should not be set.
+		Entry("Should requeue when patching an unavailable node",
+			TestCaseReconcile{
+				Objects: []runtime.Object{
+					&infrav1.BareMetalMachine{
+						TypeMeta: metav1.TypeMeta{},
```
Comment (kashifest) [inline controllers/baremetalmachine_controller_test.go]: Kind is missing?

File: controllers/baremetalmachine_controller.go [hunk 9]
```diff
@@ -201,12 +202,22 @@ func (r *BareMetalMachineReconciler) reconcileNormal(ctx context.Context,
...
+		// Set the providerID on the node if no Cloud provider
+		if machineMgr.BareMetalCluster.Spec.NoCloudProvider {
+			err = machineMgr.SetNodeProviderID(ctx, *bmhID, providerID, r.CapiClientGetter)
+			if err != nil {
+				if requeueErr, ok := errors.Cause(err).(baremetal.HasRequeueAfterError); ok {
+					return ctrl.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
+				}
+				return ctrl.Result{}, errors.Wrap(err, "failed to get the providerID for the BaremetalMachine")
+			}
+			log.Info("ProviderID set on target node")
```
Comment (kashifest) [inline controllers/baremetalmachine_controller.go]: Should we use machinelog as used in reconcile instead of log??

Comment (maelk) [reply controllers/baremetalmachine_controller.go]: it is actually the same. the log passed as parameter is machineLog in reconcile

**Pull request merged**